concurrency/mutexmap.go

```diff
@@ -58,50 +58,57 @@ func (a *mutexMap[T]) Lock(key T) {
 	a.lock.RLock()
 	mutex, ok := a.items[key]
 	a.lock.RUnlock()
+	if ok {
+		mutex.Lock()
+		return
+	}
+
+	a.lock.Lock()
+	mutex, ok = a.items[key]
 	if !ok {
-		a.lock.Lock()
-		mutex, ok = a.items[key]
-		if !ok {
-			mutex = &sync.RWMutex{}
-			a.items[key] = mutex
-		}
-		a.lock.Unlock()
+		mutex = &sync.RWMutex{}
+		a.items[key] = mutex
 	}
+	a.lock.Unlock()
 	mutex.Lock()
 }
 
 func (a *mutexMap[T]) Unlock(key T) {
 	a.lock.RLock()
 	mutex, ok := a.items[key]
-	a.lock.RUnlock()
 	if ok {
 		mutex.Unlock()
 	}
+	a.lock.RUnlock()
```

Comment:
Why is this needed? The lock of the map is now bundled with the lock of the mutex itself. The original design was keeping those separate.

Reply:
The change aims to release the outer lock after the inner operation has completed. This prevents race conditions on inner operations of the same key

Reply:
I find it odd that we need to bundle the unlock of item with the unlock of the map.

Reply:
Oh. I think I get it now. It is because of the delete.